### PR TITLE
Facing

### PR DESCRIPTION
**Pull request by KatherinaFatiguso**
Hi @itinsley 
I tried to create tests and function called  valid_place(x,y) to check if the given position is valid for placing the Robot in that place.
This valid_place will be different with valid_move(x,y) to check if the robot can move one position to its direction without falling off the table.

### CODE REVIEW

File: spec/toy_robot_spec.rb [hunk 1]
```diff
@@ -17,6 +17,36 @@
...
+    end
+
+    specify 'that when x value is not between 0 and 4 and y value is not between 0 and 4, it will return false' do
+      x = -1
+      y = 5
+      expect(subject.valid_place(x,y)).to be false
+    end
+  end ## end of describe 'valid_place'
+
+  describe 'replace_movements' do
```
Comment (itinsley) [inline spec/toy_robot_spec.rb]: replace_movements is not a public method. Make it a private method, this means you can't test it. This is one of the things I find hardest about designing interfaces and testing.
However, in this case there is no logic to worry about and the code is only called in one place so I would get rid of the method and just put the code back in the place function. 

Just test it as another assertion on the place test.

File: toy_robot.rb [hunk 2]
```diff
@@ -10,12 +10,25 @@ def initialize
...
 
+
+
   def place(x,y,facing)
-    raise InvalidPositionError.new("Invalid Position, #{x}, #{y}, #{facing}") if !(x.between?(0,4)) || !(y.between?(0,4))
-    @x = x
-    @y = y
-    @facing = facing
-    @movements.replace(["PLACE #{@x}, #{@y}, #{@facing.to_s.upcase}"])
+    if valid_place(x,y)
```
Comment (itinsley) [inline toy_robot.rb]: you won't need an else condition if you just raise the exception first.
`if !valid(x,y) raise InvalidPositionError....``

Less indenting makes the code easier to read.

**Comment by itinsley** [comment]
I don't think your assumption is correct. You should be able to use the same valid_place(x,y) condition for move. Think about  it.. all the validation we need is already in valid_place command. You just have to think about how you utilise that method.

**Comment by KatherinaFatiguso** [comment]
Hmm... I just rearrange my thinking, yes, I don't need to create valid_move, I can re-use it.

So, if facing is :north and if valid_place(x,y+1) is true, then move the Robot, right?

**Comment by KatherinaFatiguso** [comment]
Hi @itinsley 
Now that the replace_movements have been merged into place movements, then there will be the repetition of "Invalid Position, #{x}, #{y}, #{facing}" and "PLACE #{x}, #{y}, #{facing.to_s.upcase}" in one function, is that okay to do that?

And valid_place function is called in two functions: place and move, should I put valid_place into private method now?

I haven't tested in the place function if the array @movements will replace anything in it to only one item, yet.

File: spec/toy_robot_spec.rb [hunk 3]
```diff
@@ -31,48 +46,37 @@
...
-      end
-      it 'will not move from its current position' do
-        subject.place(0, 0, :south)
-        subject.move
-        expect(subject.x).to eq(0)
-        expect(subject.y).to eq(0)
-        expect(subject.facing).to eq(:south)
-      end
-    end
+    # subject { ToyRobot.new }
```
Comment (itinsley) [inline spec/toy_robot_spec.rb]: why bother commenting out code and checking it in? Keep it or delete it?

**Comment by KatherinaFatiguso** [comment]
Hi @itinsley, please check these updates.
Should I move the valid_place method as a private method?
There are so many opinions about testing private method in Stack Overflow. Some said it's not supposed to be tested, create another class using that method, use send function... what is the right thing to do?

**Comment by itinsley** [comment]
For this stuff, the private methods are very simple so just test them indirectly. You are testing  business behaviours, not functions specifically. If the function becomes sufficiently complex to justify a test, then abstract it to a separate class and test it separately. 
For the purposes of this work. Design your classes with public and private methods that make sense, then test them through the public interface.

**Comment by itinsley** [comment]
There are some bugs in your code. Try these.

  describe 'move' do
    subject { ToyRobot.new }

```
it 'When facing north will move one position north' do
  subject.place(2,2, :north)
  subject.move
  expect(subject.y).to eq(3)
end

it 'When facing south it will move one position south' do
  subject.place(2,2, :south)
  subject.move
  expect(subject.y).to eq(1)
end

it 'When facing east it will move one position east' do
  subject.place(2,2, :east)
  subject.move
  expect(subject.x).to eq(1)
end

it 'When facing west it will move one position east' do
  subject.place(2,2, :west)
  subject.move
  expect(subject.x).to eq(3)
end
```

  end ## end of describe 'move'

**Comment by itinsley** [comment]
Regarding how to validate I think it should work like this. Note, I haven't tested this code or debugged it. There is also plenty of room for a refactor once you have the tests working. Of course because you will only be testing the public interface, you can refactor private methods without having to worry about breaking tests!

In general, this PR has become very messy. If you feel you are in the weeds then you should go back to a clean state (on a different branch) and write a small piece of code with a valid test. Ensure every scenario is tested otherwise you will get false positive bugs like I think you already have in your code (see above tests).

  def move

```
# This should work but ensure you understand the difference of pass by ref or pass by value
# http://stackoverflow.com/questions/1872110/is-ruby-pass-by-reference-or-by-value

x=@x
y=@y

case @facing
when condition :north
  y += 1
when condition :south
  y += -1
when condition :east
  x += -1
when condition :west
  x += 1

if !valid_place(x,y)
  raise InvalidPositionError.new("Invalid Position, #{x}, #{y}, #{facing}")
else
  @x = x
  @y = y
  @movements << "MOVE"
end
```

  end

**Comment by KatherinaFatiguso** [comment]
Hi Ian, 
I'm trying to understand this. But why did you write this?  x=@x
Why can't you just use @x, not passing @x to x, and so the code will be shorter? 
I'm reading the notes on that Stack overflow page.
I've never used this way before.

**Comment by itinsley** [comment]
I am creating temporary variables so that we can test the placement. Before actually moving the robot, which is what we are doing when we assign to @x, @y. If you update the class variables, it's too late, the robot has fallen off the table.

File: spec/toy_robot_spec.rb [hunk 4]
```diff
@@ -31,48 +46,68 @@
...
-        subject.move
-        expect(subject.x).to eq(0)
-        expect(subject.y).to eq(1)
-        expect(subject.facing).to eq(:north)
+      it 'will move one position forward' do
+        robot = ToyRobot.new
+        robot.move
+        expect(robot.x).to eq(0)
+        expect(robot.y).to eq(1)
+        expect(robot.facing).to eq(:north)
```
Comment (itinsley) [inline spec/toy_robot_spec.rb:74]: try to avoid noise by re-testing for side effects that are unlikely. You cannot test everything. Move does not impact facing so don't re-assert it's value, moving north does not impact the X variable so don't re-assert it. Why not? Because all these extra tests, create noise that detracts from the clarity of the test. Look at my examples and they are much clearer. Set up, then test. 
You do need to test for weird side effects but they should be done with manual testing - which should test the boundaries. 
I think we can eventually refactor this to be much cleaner and test everything succinctly - but the journey is part of the process. That is the lesson.

File: spec/toy_robot_spec.rb [hunk 5]
```diff
@@ -31,48 +46,68 @@
...
   describe 'move' do
-    subject { ToyRobot.new }
     context 'When the Robot is at Origin (0, 0, :north)' do
-      it 'it will move one position forward' do
-        subject.move
-        expect(subject.x).to eq(0)
-        expect(subject.y).to eq(1)
-        expect(subject.facing).to eq(:north)
+      it 'will move one position forward' do
+        robot = ToyRobot.new
```
Comment (itinsley) [inline spec/toy_robot_spec.rb:70]: use the explicit place command as below. You are testing the #move command, not the initialiser here so be explicit. You are binding expected move behaviour to the initialiser. You should be able to change the initialiser to face a different way without breaking this test. There is no benefit to not following the pattern demonstrated below (unless i'm missing something).

**Comment by KatherinaFatiguso** [comment]
Oh, this is a good thing to learn....

I didn't think of this way before: assign the value to some temporary variables and then test it if it is valid before updating the actual class variables.

Yes, make sense... I never learnt like this in the class.

Thanks Ian

File: spec/toy_robot_spec.rb (hunk 4 above)
Comment (KatherinaFatiguso) [inline spec/toy_robot_spec.rb:74]: Ahh... I see, so, when you write a test, just care about testing what is the actual purpose of the function.

Not the other 'possible' effects, so, for these other extra things just test it manually.

This is what worries me before, because I can not imagine how many tests you have to imagine to create to test a function.

I thought I have to spend sooo much time to write all kind of possibilities that might happen of the effect of a function.

Thus I would spending too much time unnecessarily.

Thank you for clarifying this... again, never learnt this lesson from the class...

**Comment by itinsley** [comment]
this is where experience comes in. There is no exact science about testing. 
Read these: https://www.goodreads.com/author/quotes/25211.Kent_Beck
Read lots of Kent Beck: https://stanislaw.github.io/2016/01/25/notes-on-test-driven-development-by-example-by-kent-beck.html
But then you have to make your own judgements for your own code. Don't forget to do a lot of manual testing as well - that will help you discover your mistakes. When you have made one you may wish to add a test to ensure it doesn't happen again. Later on you may want to delete that test when you refactor.

**Comment by itinsley** [comment]
Regarding using temporary variables. Don't see this as having learned a trick. The lesson to take from this is that there are private variables and there are class level variables and there are publicly accessible variables (and methods). It is all about only revealing as much detail as is necessary between these 'things'. In this case you know you already have a validation method so you have to think how can i create a condition to re-use this method and save me having to write the same code again. In short, you should _never_ be repeating the same business logic rules in two places.
